TargetScale support Subcontext

falken10vdl opened the pull request:
Added the field for TargetScale in Geometric RepresentationContext
<img width="355" alt="image" src="https://github.com/user-attachments/assets/d0dd2cc1-5ce4-4caa-933a-4336ef53fb62" />

Cheers!

falken10vdl committed:
Implemented TargetScale by means of ScaleDenominator for subcontext

falken10vdl committed:
remove info in geometry module

falken10vdl committed:
black formatting

falken10vdl committed:
better naming

falken10vdl committed:
better naming

Andrej730 commented:
@falken10vdl 

From what I've seen TargetScale is used very rarily - so the most of the times users will just see `N/A` and will never use scale input when adding a new context, so it may be not worth it to introduce an unnecessary noise to the UI. 

Maybe we can add some icon before pencil button to indicate that this subcontext is special and has TargetScale set?

Regarding the TargetScale editing UI - I haven't tested it myself yet, but the idea seems reasonable - to show a denominator instead of the more obscure scale value. Can you please update tests in `tool/test_context.py` accordingly?

Regarding `context.add_context`'s new argument - please add it to the function's doc-string.

theoryshaw commented:
I could see how using `TargetScale` could open some interesting possibilities.

falken10vdl commented:
@Andrej730 and @theoryshaw what do you think about a display like this?
<img width="321" alt="image" src="https://github.com/user-attachments/assets/d50dbff8-fc7d-41a0-b7dc-2998b16dae53" />
it takes maximum 6 characters. If no scale is available then it is empty.
<img width="300" alt="image" src="https://github.com/user-attachments/assets/7cba22bd-67ac-4a4d-a9c1-5eff30993bef" />

This would be the docstring for the parameter:
<img width="624" alt="image" src="https://github.com/user-attachments/assets/374c20e6-cfe3-4a1e-a262-0676a1c3ae1e" />

tool/test_context.py.... I am not familiar with the testing environment yet...  :)   I still have to spend some time learning how it works... Anyhow I have seen that targetScale is already in one test. I have added another:

test_importing_another_subcontext_with_different_scale:
<img width="754" alt="image" src="https://github.com/user-attachments/assets/bebc2d79-c528-4477-9c2f-15727e82131a" />

test_exporting_another_subcontext_with_different_scale:
<img width="625" alt="image" src="https://github.com/user-attachments/assets/6d480c33-c05e-41bf-8144-77493be88a81" />


Let me know what you think
Cheers!

falken10vdl committed:
Updated with developers feedback

Andrej730 commented:
@falken10vdl 

> what do you think about a display like this?

Looks okay, so there won't be unnecessary "N/A" all the time. Though I still don't see the need for setting TargetScale when adding a context - it can always be set later if needed.

> tool/test_context.py.... I am not familiar with the testing environment yet... :) I still have to spend some time learning how it works... Anyhow I have seen that targetScale is already in one test. I have added another:

I believe there is no need for new tests - older import/export tests should do, they will start failing after this PR and they need to be adjusted.

falken10vdl commented:
@Andrej730 Ok. I remove the added tests since there are already two with TargetSacle being checked.

>  Though I still don't see the need for setting TargetScale when adding a context - it can always be set later if needed.

Do you mean removing this, right? and just having that in the edit menu.
<img width="517" alt="image" src="https://github.com/user-attachments/assets/9a420b8f-0baf-46a2-a445-7b2fcd9242a7" />

Thanks!

Andrej730 commented:
> Do you mean removing this, right? and just having that in the edit menu.

Yeah

falken10vdl committed:
updated per developer feedback

falken10vdl commented:
Updated. Removed redundant tests and TargetScale prop in Geometric Representation Contexts panel. TargetScale is added/changed in editing mode.
Cheers!

Andrej730 commented:
@falken10vdl thanks.

1. Regarding tests - still need to adjust existing tests for a new behaviour, I believe import and export attributes tests will fail, as they're expecting TargetScale attribute.

https://github.com/IfcOpenShell/IfcOpenShell/blob/9b58d4aa479bbde19fba7b99c7afed08286d26d9/src/bonsai/test/tool/test_context.py#L68

2. Property that was used for adding TargetScale for a new context is still present in the PR - see `bim.add_context`, `context/prop.py`, `context/ui.py`.

3. In ui.py `if "target_scale_denominator" in subcontext` seems unnecessary as all dictionaries will have this key either way.

4. In data.py `hasattr(subcontext, "TargetScale")` seems also unnecessary, see IfcGeometricRepresentationContext documentation (https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcGeometricRepresentationContext.htm) - HasSubContexts will always be a tuple of IfcGeometricRepresentationSubContext and they will always have TargetScale.

5. Please run black formatter on this PR.

falken10vdl committed:
updates + black formatting

falken10vdl commented:
@Andrej730 Thanks for the details. Please take a look to the update.
Thanks!

Andrej730 commented:
@falken10vdl 

See:

> 2\. Property that was used for adding TargetScale for a new context is still present in the PR - see `bim.add_context`, `context/prop.py`, `context/ui.py`.

falken10vdl commented:
@Andrej730 

> 2. Property that was used for adding TargetScale for a new context is still present in the PR - see bim.add_context, context/prop.py, context/ui.py.

Sorry I do not follow you :) Can you explain it?
Thanks!

Andrej730 commented:
@falken10vdl 

> Sorry I do not follow you :) Can you explain it?

Sure, `representation_target_scale_denominator` property that's not used anymore after https://github.com/IfcOpenShell/IfcOpenShell/pull/6816/commits/bb9cc11c82b5c305e6436a337d0accf1a23043f1 is still referenced by 3 files and those references can be removed to make it clean.

![image](https://github.com/user-attachments/assets/850a7981-0f1f-4db3-8259-d165a7e6ae04)
![image](https://github.com/user-attachments/assets/93359263-d47f-4e4d-a0c6-478a8f7ca5a9)
![image](https://github.com/user-attachments/assets/e00c8884-bd3b-4699-a131-06af983cf2f0)

falken10vdl committed:
removed representation_target_scale_denominator & context_target_scale_denominator as they are no longer ii the UI

falken10vdl commented:
@Andrej730 Thanks for the explanation! I had forgotten we had removed the property from the UI.
Thanks!
